internal/app/init/pkg/system/services/kubeadm/kubeadm.go

```diff
@@ -0,0 +1,304 @@
+			CertPath: constants.KubeadmFrontProxyCACert,
+			KeyPath:  constants.KubeadmFrontProxyCAKey,
+		},
+		{
+			Cert:     data.Security.Kubernetes.Etcd,
+			CertPath: constants.KubeadmEtcdCACert,
+			KeyPath:  constants.KubeadmEtcdCAKey,
+		},
+	}
+
+	for _, cert := range certs {
+		if cert.Cert == nil {
+			continue
+		}
+
+		if cert.Cert.Crt != nil {
+			if err = os.MkdirAll(path.Dir(cert.CertPath), dirPerm); err != nil {
+				return err
+			}
+			if err = ioutil.WriteFile(cert.CertPath, cert.Cert.Crt, certPerm); err != nil {
+				return fmt.Errorf("write %s: %v", cert.CertPath, err)
+			}
+		}
+		if cert.Cert.Key != nil {
+			if err = os.MkdirAll(path.Dir(cert.KeyPath), dirPerm); err != nil {
+				return err
+			}
+			if err = ioutil.WriteFile(cert.KeyPath, cert.Cert.Key, certPerm); err != nil {
+				return fmt.Errorf("write %s: %v", cert.KeyPath, err)
+			}
+		}
+	}
+	return err
+}
+
+// RequiredFiles returns a slice of the required CA and
+// security policies necessary for kubeadm init to function.
+// This serves as a base for the list of files that need to
+// be synced via trustd from other nodes
+func RequiredFiles() []string {
+	return []string{
+		constants.AuditPolicyPathInitramfs,
+		constants.EncryptionConfigInitramfsPath,
+		constants.KubeadmCACert,
+		constants.KubeadmCAKey,
+		constants.KubeadmSACert,
+		constants.KubeadmSAKey,
+		constants.KubeadmFrontProxyCACert,
+		constants.KubeadmFrontProxyCAKey,
+		constants.KubeadmEtcdCACert,
+		constants.KubeadmEtcdCAKey,
+	}
+}
+
+// FileSet compares the list of required files to the ones
+// already present on the node and returns the delta
+func FileSet() []*proto.ReadFileRequest {
+	files := RequiredFiles()
+
+	fileRequests := []*proto.ReadFileRequest{}
+	// Check to see if we already have the file locally
+	for _, file := range files {
+		if _, err := os.Stat(file); os.IsNotExist(err) {
+			fileRequests = append(fileRequests, &proto.ReadFileRequest{Path: file})
+		}
+	}
+
+	return fileRequests
+}
+
+// CreateTrustdClients handles instantiating a trustd client connection
+// to each trustd endpoint defined in userdata
+func CreateTrustdClients(data *userdata.UserData) ([]proto.TrustdClient, error) {
+	var creds basic.Credentials
+	var err error
+
+	trustds := []proto.TrustdClient{}
+
+	creds, err = basic.NewCredentials(data.Services.Trustd)
+	if err != nil {
+		return trustds, err
+	}
+
+	// Create a trustd client for each endpoint to set up
+	// a fan out approach to gathering the files
+	var conn *grpc.ClientConn
+	for _, endpoint := range data.Services.Trustd.Endpoints {
+		conn, err = basic.NewConnection(endpoint, constants.TrustdPort, creds)
+		if err != nil {
+			return trustds, err
+		}
+		trustds = append(trustds, proto.NewTrustdClient(conn))
+	}
+	return trustds, nil
+}
+
+// Download handles the retrieval of files from a trustd endpoint
+func Download(ctx context.Context, client proto.TrustdClient, file *proto.ReadFileRequest, content chan<- []byte) {
+	select {
```

Comment:
idk if it deserves a comment, but this function does very important thing: if some of the calls to trustd succeeds and delivers something to `content` channel, main loop will read the value, cancel the context, and it will never read from the channel `content`. so we need to push to `content` in a `select` like here to make sure we're not stuck pushing to a channel nobody is reading from

Reply:
Wouldn't we read from the ctx.Done() chan in `Download()` and prevent that from happening? 
I'm under the impression select will pick the first option that won't block, so if we have 1x Download() func being called per trustd endpoint, the first one read would win, be used, and then trigger ctxCancel(). This should cause the select statement to read from ctx.Done() and no longer block trying to send to the content channel.

Reply:
yes, exactly, my point is that probably it's not obvious (and might deserve some comment for future developers)